### PR TITLE
feat: add GRPO trainer to mlx-lm tuner

### LINKED ISSUE #1420: Add GRPO trainer
RT. Just to add support for more RL use cases.
Should it be part of `tuner/`?

### PR DESCRIPTION
**Pull request by HaoXuAI**
## Summary

First RL training capability in mlx-lm: **GRPO (Group Relative Policy Optimization)** trainer.

Generates K completions per prompt, scores them with a user-provided reward function, computes group-normalized advantages, and applies a REINFORCE-style policy gradient with LoRA.

## API

```python
from mlx_lm import load
from mlx_lm.tuner import linear_to_lora_layers
from mlx_lm.tuner.grpo_trainer import GRPOArgs, grpo_train

model, tokenizer = load('mlx-community/Qwen2.5-0.5B-Instruct-4bit')
linear_to_lora_layers(model, num_layers=16, lora_config={'rank': 8})

def reward_fn(completions, prompt):
    return [1.0 if 'correct' in c else 0.0 for c in completions]

grpo_train(model, tokenizer, prompts=['Solve 2+2'], reward_fn=reward_fn,
           args=GRPOArgs(num_completions=8, max_new_tokens=256))
```

## How it works

Each iteration:
1. Generate K completions from current policy
2. Score with user-provided `reward_fn(completions, prompt) -> List[float]`
3. Compute group-normalized advantages: `(r - mean) / std`
4. REINFORCE loss: `-mean(advantage * log_prob_completion)`
5. Gradient clip + optimizer step
6. Skip steps where all rewards are identical (no gradient signal)

## Motivation

With DeepSeek-R1 and Qwen3 demonstrating the power of RL for reasoning, there's strong demand for local GRPO training on Apple Silicon. Currently mlx-lm supports SFT and DPO (offline preference) but has no on-policy RL capability.

This trainer enables:
- Local RLHF/GRPO with any mlx-lm model + LoRA
- Pluggable reward functions (rule-based, model-based, tool-verified)
- Used in practice for Kaggle competition training (Nemotron-H 30B, 4-bit)

## Changes

- `mlx_lm/tuner/grpo_trainer.py`: ~280 lines
  - `GRPOArgs` dataclass (hyperparameters)
  - `GRPOCallback` class (monitoring hooks)
  - `grpo_train()` function (main entry point)
- `tests/test_grpo_trainer.py`: 3 tests

## Related

- Batched same-prompt generation for speedup: #1403
- Per-token logprobs from batch_generate: #1358 (merged)

fix: https://github.com/ml-explore/mlx-lm/issues/1420

### CODE REVIEW

**Comment by angeloskath** [comment]
I 'll close this in favor of https://github.com/Goekdeniz-Guelmez/mlx-lm-lora but if you think that something is not covered there or there is a specific benefit to having in MLX-LM feel free to reopen.
